Add remaining fields to TPU VM resource

roaks3 opened the pull request:
b/292585830

This adds the remaining fields to `google_tpu_v2_vm` that are available from the API.

Notable exceptions (which may end up making it into this PR):
- `accelerator_config`
    - This conflicts with `accelerator_type`, and we'll need to work around the existing default behavior

One other interesting piece is `service_account.scopes`. If the user specifies this field but does not include `https://www.googleapis.com/auth/pubsub`, it will be automatically added by the server, so I've added a diff_suppress to ignore that value.

<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->


<!--
Please self-review your PR against the review checklist before creating it: https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/

Completing the checklist will help speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.

If your PR is still work in progress, please create it in draft mode
-->

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none

Unless you choose release-note:none, please add a release note.

See https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/ for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
tpuv2: added more fields to `google_tpu_v2_vm` resource, including `network_config`, `scheduling_config`, `shielded_instance_config`, `service_account` and `data_disks`
```

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 218 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9308-old..auto-pr-9308) ( 4 files changed, 1247 insertions(+), 34 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 268 insertions(+))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 52 insertions(+))
## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_tpu_v2_vm` (5 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_tpu_v2_vm" "primary" {
  scheduling_config {
    reserved = # value needed
  }
}

```

modular-magician commented:
#### Tests analytics 
Total tests: `3162` 
Passed tests `2840` 
Skipped tests: `318` 
Affected tests: `4` 

#### Action taken 
<details> <summary>Found 4 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccContainerAttachedCluster_containerAttachedClusterFullExample|TestAccContainerAttachedCluster_update|TestAccTpuV2Vm_update|TestAccTpuV2Vm_tpuV2VmFullExample </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccContainerAttachedCluster_containerAttachedClusterFullExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/1f367723-13d0-4c0d-9569-3e0de3b2af45/recording/TestAccContainerAttachedCluster_containerAttachedClusterFullExample.log)]
`TestAccContainerAttachedCluster_update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/1f367723-13d0-4c0d-9569-3e0de3b2af45/recording/TestAccContainerAttachedCluster_update.log)]
`TestAccTpuV2Vm_tpuV2VmFullExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/1f367723-13d0-4c0d-9569-3e0de3b2af45/recording/TestAccTpuV2Vm_tpuV2VmFullExample.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{red}{\textsf{Tests failed during RECORDING mode:}}$ 
 `TestAccTpuV2Vm_update`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/1f367723-13d0-4c0d-9569-3e0de3b2af45/build-log/recording_build/TestAccTpuV2Vm_update_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/1f367723-13d0-4c0d-9569-3e0de3b2af45/recording/TestAccTpuV2Vm_update.log)] 

$\textcolor{red}{\textsf{Please fix these to complete your PR.}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/1f367723-13d0-4c0d-9569-3e0de3b2af45/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/1f367723-13d0-4c0d-9569-3e0de3b2af45/recording) for each test

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 218 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9308-old..auto-pr-9308) ( 4 files changed, 1255 insertions(+), 34 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 268 insertions(+))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 52 insertions(+))
## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_tpu_v2_vm` (5 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_tpu_v2_vm" "primary" {
  scheduling_config {
    reserved = # value needed
  }
}

```

modular-magician commented:
#### Tests analytics 
Total tests: `3162` 
Passed tests `2841` 
Skipped tests: `318` 
Affected tests: `3` 

#### Action taken 
<details> <summary>Found 3 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccContainerAttachedCluster_containerAttachedClusterFullExample|TestAccContainerAttachedCluster_update|TestAccTpuV2Vm_update </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccContainerAttachedCluster_containerAttachedClusterFullExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/ab229bbf-582a-4746-8087-5c3dc5069c39/recording/TestAccContainerAttachedCluster_containerAttachedClusterFullExample.log)]
`TestAccContainerAttachedCluster_update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/ab229bbf-582a-4746-8087-5c3dc5069c39/recording/TestAccContainerAttachedCluster_update.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{red}{\textsf{Tests failed during RECORDING mode:}}$ 
 `TestAccTpuV2Vm_update`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/ab229bbf-582a-4746-8087-5c3dc5069c39/build-log/recording_build/TestAccTpuV2Vm_update_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/ab229bbf-582a-4746-8087-5c3dc5069c39/recording/TestAccTpuV2Vm_update.log)] 

$\textcolor{red}{\textsf{Please fix these to complete your PR.}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/ab229bbf-582a-4746-8087-5c3dc5069c39/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/ab229bbf-582a-4746-8087-5c3dc5069c39/recording) for each test

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 218 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9308-old..auto-pr-9308) ( 4 files changed, 1263 insertions(+), 32 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 268 insertions(+))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 52 insertions(+))
## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_tpu_v2_vm` (5 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_tpu_v2_vm" "primary" {
  scheduling_config {
    reserved = # value needed
  }
}

```

modular-magician commented:
#### Tests analytics 
Total tests: `3164` 
Passed tests `2840` 
Skipped tests: `318` 
Affected tests: `6` 

#### Action taken 
<details> <summary>Found 6 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccContainerNodePool_fastSocket|TestAccContainerNodePool_012_ConfigModeAttr|TestAccDataSourceGoogleServiceAccountJwt|TestAccDataSourceGoogleServiceAccountAccessToken_basic|TestAccTpuV2Vm_update|TestAccWorkstationsWorkstationConfig_workstationConfigAcceleratorsExample </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

roaks3 commented:
I've added you both for initial review since this should be done (except for the additional fields I will be trying to add), and tests have passed locally, but it is just taking a very long time to get the VCR tests in a happy place.

I already have a change queued up to attempt adding the networking fields as well, just waiting on VCR to complete to do that.

@rileykarson I added you since I know you have context but feel free to re-assign.

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccContainerNodePool_012_ConfigModeAttr`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/5fbf2526-ab48-4851-abfc-6be17f85136a/recording/TestAccContainerNodePool_012_ConfigModeAttr.log)]
`TestAccDataSourceGoogleServiceAccountAccessToken_basic`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/5fbf2526-ab48-4851-abfc-6be17f85136a/recording/TestAccDataSourceGoogleServiceAccountAccessToken_basic.log)]
`TestAccTpuV2Vm_update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/5fbf2526-ab48-4851-abfc-6be17f85136a/recording/TestAccTpuV2Vm_update.log)]
`TestAccWorkstationsWorkstationConfig_workstationConfigAcceleratorsExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/5fbf2526-ab48-4851-abfc-6be17f85136a/recording/TestAccWorkstationsWorkstationConfig_workstationConfigAcceleratorsExample.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{red}{\textsf{Tests failed during RECORDING mode:}}$ 
 `TestAccContainerNodePool_fastSocket`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/5fbf2526-ab48-4851-abfc-6be17f85136a/build-log/recording_build/TestAccContainerNodePool_fastSocket_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/5fbf2526-ab48-4851-abfc-6be17f85136a/recording/TestAccContainerNodePool_fastSocket.log)]
`TestAccDataSourceGoogleServiceAccountJwt`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/5fbf2526-ab48-4851-abfc-6be17f85136a/build-log/recording_build/TestAccDataSourceGoogleServiceAccountJwt_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/5fbf2526-ab48-4851-abfc-6be17f85136a/recording/TestAccDataSourceGoogleServiceAccountJwt.log)] 

$\textcolor{red}{\textsf{Please fix these to complete your PR.}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/5fbf2526-ab48-4851-abfc-6be17f85136a/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/5fbf2526-ab48-4851-abfc-6be17f85136a/recording) for each test

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 257 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9308-old..auto-pr-9308) ( 4 files changed, 1410 insertions(+), 32 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 300 insertions(+))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 72 insertions(+))
## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_tpu_v2_vm` (5 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_tpu_v2_vm" "primary" {
  scheduling_config {
    reserved = # value needed
  }
}

```

modular-magician commented:
#### Tests analytics 
Total tests: `3163` 
Passed tests `2841` 
Skipped tests: `317` 
Affected tests: `5` 

#### Action taken 
<details> <summary>Found 5 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccContainerNodePool_012_ConfigModeAttr|TestAccProjectIamPolicy_invalidMembers|TestAccTpuV2Vm_tpuV2VmFullExample|TestAccTpuV2Vm_update|TestAccWorkstationsWorkstationConfig_workstationConfigAcceleratorsExample </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

rileykarson reviewed (commented):
Partial pass, I'm still working through updatable fields.

rileykarson commented on mmv1/products/tpuv2/Vm.yaml:
```diff
@@ -88,3 +94,220 @@ properties:
+          Allows the TPU node to send and receive packets with non-matching destination or source
+          IPs. This is required if you plan to use the TPU workers to forward routes.
+  - !ruby/object:Api::Type::NestedObject
+    name: 'serviceAccount'
```
I think we can safely mark this immutable- it can still safely be removed from config because it's `default_from_api`, and any other change will recreate it.

roaks3 replied:
Will do!

rileykarson commented on mmv1/third_party/terraform/services/tpuv2/resource_tpu_v2_vm_test.go.erb:
```diff
@@ -35,63 +30,134 @@ func TestAccTpuV2Vm_update(t *testing.T) {
+    preemptible = true
+  }
+
+  data_disks {
```
Can we add a disk present in both configs that we flip from one `mode` to another?

roaks3 replied:
I initially tried that but wasn't able to get a test working with the `READ_WRITE` value on a disk I created:
```
Error 400: Cloud TPU received a bad request. Requested data disk is invalid: disk "projects/terraform-dev-ryanoaks/zones/us-central1-a/disks/tf-test-tpu2-gme8kmr81r" is invalid: generic::invalid_argument: User specified attached disks that are shared across a pod must be in 'READ ONLY' mode, but disk was in mode 
```

rileykarson replied:
Weird! We should be working with a single node here if I understand the API correctly.

roaks3 replied:
This actually worked when I tried again, so it is included now. I suspect something about my config was making the `READ_WRITE` option unavailable, but not sure what specifically 🤷

rileykarson reviewed (commented):
OK, pass done- was shorter to finish than I expected.

roaks3 commented on mmv1/products/tpuv2/Vm.yaml:
```diff
@@ -88,3 +94,220 @@ properties:
+      Network configurations for the TPU node.
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'network'
```
I'm not sure, I can check

roaks3 replied:
Ok, I confirmed that you can specify `subnetwork` without `network`. Is there a config change you were thinking of that would capture that?

rileykarson replied:
Assuming you tested with a non-default subnetwork we may want to change the docs, since I'm assuming the field is either redundant or defaults to the network of the subnetwork in that case.

roaks3 replied:
Ah yea, I see some other resources have better descriptions, which appear to better match the behavior I'm seeing. What if we changed to:
- network: "The name of the network for the TPU node. It must be a preexisting Google Compute Engine network. If both network and subnetwork are specified, the given subnetwork must belong to the given network. If network is not specified, it will be looked up from the subnetwork if one is provided, or otherwise use "default"."
- subnetwork: "The name of the subnetwork for the TPU node. It must be a preexisting Google Compute Engine subnetwork. If both network and subnetwork are specified, the given subnetwork must belong to the given network. If subnetwork is not specified, the subnetwork with the same name as the network will be used."

rileykarson replied:
Assuming that's consistent with the behaviour of the resource, yep!

roaks3 replied:
It is! I tested out the scenarios described to confirm.

roaks3 commented on mmv1/products/tpuv2/Vm.yaml:
```diff
@@ -88,3 +94,220 @@ properties:
+          The name of the subnetwork for the TPU node. It must be a preexisting Google Compute
+          Engine subnetwork. If none is provided, "default" will be used.
+      - !ruby/object:Api::Type::Boolean
+        name: 'enableExternalIps'
```
Ah ok, I think we'll want the `default_value` then. I had originally added the `default_from_api` based on this diff when not specifying `network_config`:
```
             - network_config {
                  - can_ip_forward      = false -> null
                  - enable_external_ips = false -> null
                  - network             = "projects/terraform-dev-ryanoaks/global/networks/default" -> null
                  - subnetwork          = "projects/terraform-dev-ryanoaks/regions/us-central1/subnetworks/default" -> null
                }
```

rileykarson replied:
Ah- no `default_value` needed, since `false` is the default default for booleans. We should be able to delete `default_from_api`.

That was happening because the block wasn't specified, and Terraform doesn't supported nested objects natively- making the block O+C fixes that. A configuration of `network_config {}` would show the following diff without `default_from_api` on any subelements:

```
network_config {
                  - network             = "projects/terraform-dev-ryanoaks/global/networks/default" -> null
                  - subnetwork          = "projects/terraform-dev-ryanoaks/regions/us-central1/subnetworks/default" -> null
   }
```

(that's why network/subnetwork need to keep `default_from_api`, they have a complex default)

roaks3 replied:
Makes sense! I didn't have a strong understanding of some of these configs so thanks for explaining.

roaks3 commented on mmv1/products/tpuv2/Vm.yaml:
```diff
@@ -88,3 +94,220 @@ properties:
+          Allows the TPU node to send and receive packets with non-matching destination or source
+          IPs. This is required if you plan to use the TPU workers to forward routes.
+  - !ruby/object:Api::Type::NestedObject
+    name: 'serviceAccount'
```
I don't think it's meaningful to specify `service_account {}` (maybe to re-create with default values?), but if I recall, I modeled it this way because you can specify `email` or `scope` without the other being required.

rileykarson replied:
Got it! Should be fine as-is then.

roaks3 commented on mmv1/products/tpuv2/Vm.yaml:
```diff
@@ -88,3 +94,220 @@ properties:
     description: |
       Text description of the TPU.
+  - !ruby/object:Api::Type::String
+    name: 'cidrBlock'
```
I actually had a note for this field to check that after tests pass. I haven't been able to verify so far, but agreed that this probably isn't mutable.

roaks3 replied:
Yep, confirmed it isn't mutable.

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccContainerNodePool_012_ConfigModeAttr`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/51392b3f-1ea9-4773-ad4f-579235302e06/recording/TestAccContainerNodePool_012_ConfigModeAttr.log)]
`TestAccProjectIamPolicy_invalidMembers`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/51392b3f-1ea9-4773-ad4f-579235302e06/recording/TestAccProjectIamPolicy_invalidMembers.log)]
`TestAccTpuV2Vm_tpuV2VmFullExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/51392b3f-1ea9-4773-ad4f-579235302e06/recording/TestAccTpuV2Vm_tpuV2VmFullExample.log)]
`TestAccTpuV2Vm_update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/51392b3f-1ea9-4773-ad4f-579235302e06/recording/TestAccTpuV2Vm_update.log)]
`TestAccWorkstationsWorkstationConfig_workstationConfigAcceleratorsExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/51392b3f-1ea9-4773-ad4f-579235302e06/recording/TestAccWorkstationsWorkstationConfig_workstationConfigAcceleratorsExample.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{green}{\textsf{All tests passed!}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/51392b3f-1ea9-4773-ad4f-579235302e06/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/51392b3f-1ea9-4773-ad4f-579235302e06/recording) for each test

rileykarson reviewed (approved):
LGTM, nothing blocking left on my end- it'd be nice to test updating a disk mode too, but as you mentioned that wasn't succeeding for API business logic reasons.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 258 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9308-old..auto-pr-9308) ( 4 files changed, 1399 insertions(+), 32 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 302 insertions(+))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 72 insertions(+))
## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_tpu_v2_vm` (5 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_tpu_v2_vm" "primary" {
  scheduling_config {
    reserved = # value needed
  }
}

```

modular-magician commented:
#### Tests analytics 
Total tests: `3163` 
Passed tests `2844` 
Skipped tests: `317` 
Affected tests: `2` 

#### Action taken 
<details> <summary>Found 2 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccBigQueryDataTable_bigtable|TestAccTpuV2Vm_update </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccBigQueryDataTable_bigtable`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/cb65a19c-7a76-4458-a60d-0dbf1020d341/recording/TestAccBigQueryDataTable_bigtable.log)]
`TestAccTpuV2Vm_update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/cb65a19c-7a76-4458-a60d-0dbf1020d341/recording/TestAccTpuV2Vm_update.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{green}{\textsf{All tests passed!}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/cb65a19c-7a76-4458-a60d-0dbf1020d341/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/cb65a19c-7a76-4458-a60d-0dbf1020d341/recording) for each test

zeleena commented on mmv1/templates/terraform/examples/tpu_v2_vm_full.tf.erb:
```diff
@@ -15,4 +15,76 @@ resource "google_tpu_v2_vm" "<%= ctx[:primary_resource_id] %>" {
+  provider = google-beta
+
+  account_id   = "<%= ctx[:vars]['sa_id'] %>"
+  display_name = "Test TPU V2 VM"
```
Externally we call this a "TPU VM" whereas v1 is a "TPU Node"

zeleena commented on mmv1/third_party/terraform/services/tpuv2/resource_tpu_v2_vm_test.go.erb:
```diff
@@ -35,63 +30,139 @@ func TestAccTpuV2Vm_update(t *testing.T) {
+  tags = ["foo"]
+
+  lifecycle {
+    prevent_destroy = true
```
Why do we want prevent_destroy to be true?

roaks3 replied:
This is sort of a trick I've seen used a few times that might be adopted more. The point of this type of test is to create the resource, update it in-place, and then destroy it. However, by default the "update in-place" requirement is not enforced, so changing an immutable field will still pass, and since that will recreate the resource, you can no longer tell if _any_ of the fields are appropriately configured in terms of mutability.

So the workaround you see here ensures that Terraform will error if the resource needs to be recreated, so you know that only in-place updates are taking place if the test passes. Then, in order to clean up after the test is done, we need to set `prevent_destroy = false`, and then finally delete the resource.

rileykarson replied:
Longer term we'll be able to use https://github.com/hashicorp/terraform-plugin-testing/issues/61 to assert properties of the test plan

zeleena commented on mmv1/third_party/terraform/services/tpuv2/resource_tpu_v2_vm_test.go.erb:
```diff
@@ -35,63 +30,139 @@ func TestAccTpuV2Vm_update(t *testing.T) {
+resource "google_compute_disk" "disk" {
   provider = google-beta
+
+  name  = "tf-test-tpu-%{random_suffix}"
```
nit: can we give these disks better names? I was thinking it might be misleading when debugging tests if it's called "tpu" rather than something with "disk"

roaks3 replied:
This choice is largely from the perspective of our team seeing resources accumulating in our nightly test projects, and that may not match typical usage. But basically, when a bunch of disks are left in the instance, it is typically easier to resolve if we can identify which test the resource came from (ie. a tpu test).

Happy to change this to align to other usage though. Maybe `tf-test-tpu-disk-` is fine?

zeleena replied:
`tf-test-tpu-disk-` sounds like a good compromise to me

roaks3 commented on mmv1/products/tpuv2/Vm.yaml:
```diff
@@ -88,3 +94,220 @@ properties:
+    item_type: Api::Type::String
+  - !ruby/object:Api::Type::String
+    name: 'state'
+    output: true
```
Yep, `output: true` is our equivalent to output-only; it cannot be set by the user. And yea, this is where we specify that for mmv1 resources.

zeleena replied:
I see. It didn't seem required to state the outputs when I tested it and I didn't see this in some of the other resource files I checked. This seems fine, just curious if this is the preferred method since I'm not sure if I saw any difference when output fields were omitted.

roaks3 replied:
Yea, you won't typically see a difference in output fields when testing, but we include them to better support users. They can feed outputs into other config or automation they may be using.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 262 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9308-old..auto-pr-9308) ( 4 files changed, 1407 insertions(+), 32 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 302 insertions(+))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 72 insertions(+))
## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_tpu_v2_vm` (5 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_tpu_v2_vm" "primary" {
  scheduling_config {
    reserved = # value needed
  }
}

```

modular-magician commented:
#### Tests analytics 
Total tests: `3164` 
Passed tests `2845` 
Skipped tests: `317` 
Affected tests: `2` 

#### Action taken 
<details> <summary>Found 2 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccSpannerInstanceIamPolicy|TestAccTpuV2Vm_tpuV2VmFullExample </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccSpannerInstanceIamPolicy`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/b10923a3-2ad5-4833-9320-2b595652445f/recording/TestAccSpannerInstanceIamPolicy.log)]
`TestAccTpuV2Vm_tpuV2VmFullExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/b10923a3-2ad5-4833-9320-2b595652445f/recording/TestAccTpuV2Vm_tpuV2VmFullExample.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{green}{\textsf{All tests passed!}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/b10923a3-2ad5-4833-9320-2b595652445f/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/b10923a3-2ad5-4833-9320-2b595652445f/recording) for each test

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 262 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9308-old..auto-pr-9308) ( 4 files changed, 1407 insertions(+), 32 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 302 insertions(+))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 72 insertions(+))
## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_tpu_v2_vm` (5 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_tpu_v2_vm" "primary" {
  scheduling_config {
    reserved = # value needed
  }
}

```

modular-magician commented:
#### Tests analytics 
Total tests: `3170` 
Passed tests `2849` 
Skipped tests: `320` 
Affected tests: `1` 

#### Action taken 
<details> <summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccTpuV2Vm_update </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccTpuV2Vm_update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/bb15360b-2ae9-43a6-8fcc-1d376007180c/recording/TestAccTpuV2Vm_update.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{green}{\textsf{All tests passed!}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/bb15360b-2ae9-43a6-8fcc-1d376007180c/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/bb15360b-2ae9-43a6-8fcc-1d376007180c/recording) for each test

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 272 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9308-old..auto-pr-9308) ( 4 files changed, 1575 insertions(+), 74 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 302 insertions(+))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-9308-old..auto-pr-9308) ( 1 file changed, 72 insertions(+))
## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_tpu_v2_vm` (5 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_tpu_v2_vm" "primary" {
  scheduling_config {
    reserved = # value needed
  }
}

```

modular-magician commented:
#### Tests analytics 
Total tests: `3170` 
Passed tests `2850` 
Skipped tests: `320` 
Affected tests: `0` 

$\textcolor{green}{\textsf{All tests passed in REPLAYING mode.}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9308/artifacts/2dd5d5b0-6a7f-48bf-8d1a-6654082baeaf/build-log/replaying_test.log)
